modules/miner/blockmanager.go

```diff
@@ -69,21 +69,53 @@ func (m *Miner) HeaderForWork() (types.BlockHeader, types.Target) {
 	lockID := m.mu.Lock()
 	defer m.mu.Unlock(lockID)
 
-	// Grab a block for work.
-	block, target := m.blockForWork()
+	var header types.BlockHeader
+	var randTxn types.Transaction
+	var block *types.Block
 
-	// Save a mapping between the block and its header, replacing the block
-	// that was stored 'headerForWorkMemory' requests ago.
+	if m.memProgress%headersPerBlockMemory == 0 {
+		// Grab a new block
+		block = new(types.Block)
+		*block, _ = m.blockForWork()
+		header = block.Header()
+		randTxn = block.Transactions[0]
+	} else {
+		// Set block to previous block and create a randTxn
+		block = m.blockMem[m.headerMem[m.memProgress-1]]
+
+		randBytes, _ := crypto.RandBytes(types.SpecifierLen)
+		randTxn = types.Transaction{
+			ArbitraryData: [][]byte{append(modules.PrefixNonSia[:], randBytes...)},
+		}
+
+		// Overwrite the old block's random transaction
+		blockTransactions := append([]types.Transaction{randTxn}, block.Transactions[1:]...)
```

Comment:
There's an easier way to do this:

`block.Transactions[0].ArbitraryData[0] = randBytes`

Then you don't have to make a transaction or do the (somewhat expensive) append call.